src/tests/suite/daos_checksum.c

```diff
@@ -0,0 +1,185 @@
+{
+	int rc;
+
+	/** close object */
+	rc = daos_obj_close(ctx->oh, NULL);
+	assert_int_equal(rc, 0);
+
+	/** Close & Destroy Container */
+	rc = daos_cont_close(ctx->coh, NULL);
+	assert_int_equal(rc, 0);
+	rc = daos_cont_destroy(ctx->poh, ctx->uuid, true, NULL);
+	assert_int_equal(rc, 0);
+
+	d_sgl_fini(&ctx->sgl, true);
+}
+
+static void
+io_with_server_side_verify(void **state)
+{
+	struct csum_test_context	 ctx = {0};
+	daos_prop_t			*props;
+	int				 rc;
+
+	/**
+	 * Setup
+	 */
+	ctx.poh = ((test_arg_t *)*state)->pool.poh;
+
+	/** Container props for checksum (default chunksize is good) */
+	props = daos_prop_alloc(2);
+	props->dpp_entries[0].dpe_type = DAOS_PROP_CO_CSUM;
+	props->dpp_entries[0].dpe_val = DAOS_PROP_CO_CSUM_CRC64;
+	props->dpp_entries[1].dpe_type = DAOS_PROP_CO_CSUM_SERVER_VERIFY;
+	props->dpp_entries[1].dpe_val = DAOS_PROP_CO_CSUM_SV_OFF;
+
+	/**
+	 * Act - testing four use cases
+	 * 1. Regular, server verify disabled and no corruption ... obviously
+	 *    should be success.
+	 * 2. Server verify enabled, and still no corruption. Should be success.
+	 * 3. Server verify disabled and there's corruption. Update should
+	 *    still be success because the corruption won't be caught until
+	 *    it's fetched.
+	 * 4. Server verify enabled and corruption occurs. The update should
+	 *    fail because the server will catch the corruption.
+	 *
+	 */
+	/** 1. Server verify disabled, no corruption */
+	setup_test_context(&ctx, props);
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, &ctx.dkey, 1,
+		&ctx.iod, &ctx.sgl, NULL);
+	assert_int_equal(rc, 0);
+	cleanup_test(&ctx);
+
+	/** 2. Server verify enabled, no corruption */
+	props->dpp_entries[1].dpe_val = DAOS_PROP_CO_CSUM_SV_ON;
+	setup_test_context(&ctx, props);
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, &ctx.dkey, 1,
+			     &ctx.iod, &ctx.sgl, NULL);
+	assert_int_equal(rc, 0);
+	cleanup_test(&ctx);
+
+	/** 3. Server verify disabled, corruption occurs, update should work */
+	props->dpp_entries[1].dpe_val = DAOS_PROP_CO_CSUM_SV_OFF;
+	daos_fail_loc_set(DAOS_CHECKSUM_UPDATE_FAIL | DAOS_FAIL_ALWAYS);
+	setup_test_context(&ctx, props);
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, &ctx.dkey, 1,
+			     &ctx.iod, &ctx.sgl, NULL);
+	assert_int_equal(rc, 0);
+	cleanup_test(&ctx);
+
+	/** 4. Server verify enabled, corruption occurs, update should fail */
+	props->dpp_entries[1].dpe_val = DAOS_PROP_CO_CSUM_SV_ON;
+	setup_test_context(&ctx, props);
+	rc = daos_obj_update(ctx.oh, DAOS_TX_NONE, &ctx.dkey, 1,
+			     &ctx.iod, &ctx.sgl, NULL);
+	assert_int_equal(rc, -DER_IO);
+	cleanup_test(&ctx);
+
+	/**
+	 * Clean Up
+	 */
+	daos_prop_free(props);
+}
+
+static int
+setup(void **state)
+{
+	return test_setup(state, SETUP_POOL_CONNECT, true, DEFAULT_POOL_SIZE,
+			  NULL);
+}
+
+static const struct CMUnitTest tests[] = {
+	{ "DAOS_CSUM01: simple update with server side verify",
+		io_with_server_side_verify, async_disable, test_case_teardown},
+
```

Comment:
I guess you will add more test cases in the future, right?
For example, for the cases that:
A). There is data corruption when transferred from server to client (for fetch case).
B). There is on disk data corruption, so the local fetch on server will detect inconsistency or until client found that as case A).

Reply:
Correct. Checksums in the fetch path aren't implemented yet. Working on it now actually in another change. With it will come more tests.